### PR TITLE
 [Dubbo-3829] support google pb generic invocation

### PR DESCRIPTION
**Pull request by vio-lin**
## What is the purpose of the change
Currently, Generic Reference cannot support Google Protobuf. With bride usage of Google PB, we are going to support it via Generic Reference. See issue #3829

## Brief changelog
1.  add ProtobufUtils.class  to serialize /deserialize Google Protobuf Object
2.  modify GenericFilter.class and ProtocolUtils.class

## feedback wanted
To support Google PB Testing, GenericFilter is part of my change, also Service Metadata have to support that. If the solution is ok, i will continue on the Service Metadata part.

- [x] Make sure there is a GITHUB_issue field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x]   Format the pull request title like [Dubbo-XXX] Fix UnknownException when host config not exist #XXX. Each commit in the pull request should have a meaningful subject line and body.
- [x]   Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x]   Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in test module.
- [x]   Run mvn clean install -DskipTests=false & mvn clean test-compile failsafe:integration-test to make sure unit-test and integration-test pass.
- [x]   If this contribution is large, please follow the Software Donation Guide.

### CODE REVIEW

**Comment by beiwei30** [comment]
I revisit this change today. I am hesitate to introduce protobuf dependency in dubbo-rpc-api too. Is it possible to introduce protobuf serialization impl in 'dubbo-serialization' first, then use it in GenericFilter like this:

```java
                if (ProtocolUtils.isJavaGenericSerialization(generic)) {
                    try {
                        UnsafeByteArrayOutputStream os = new UnsafeByteArrayOutputStream(512);
                        ExtensionLoader.getExtensionLoader(Serialization.class)
                                .getExtension(Constants.GENERIC_SERIALIZATION_NATIVE_JAVA)
                                .serialize(null, os).writeObject(result.getValue());

````

or does 'dubbo-protostuff-serialization' meet your needs?

File: dubbo-all/pom.xml [hunk 1]
```diff
@@ -387,6 +387,13 @@
             <scope>compile</scope>
             <optional>true</optional>
         </dependency>
+        <dependency>
+            <groupId>org.apache.dubbo</groupId>
+            <artifactId>dubbo-serialization-googlePb</artifactId>
```
Comment (beiwei30) [inline dubbo-all/pom.xml]: let's call it 'dubbo-serialization-protobuf-json'

File: dubbo-bom/pom.xml [hunk 2]
```diff
@@ -342,6 +342,11 @@
                 <artifactId>dubbo-serialization-gson</artifactId>
                 <version>${project.version}</version>
             </dependency>
+            <dependency>
+                <groupId>org.apache.dubbo</groupId>
+                <artifactId>dubbo-serialization-googlePb</artifactId>
```
Comment (beiwei30) [inline dubbo-bom/pom.xml]: rename to 'dubbo-serialization-protobuf-json'

File: dubbo-common/src/main/java/org/apache/dubbo/common/Constants.java [hunk 3]
```diff
@@ -727,6 +727,8 @@
 
     public static final String GENERIC_SERIALIZATION_BEAN = "bean";
 
+    public static final String GENERIC_SERIALIZATION_PROTOBUF = "gernericprotobuf";
```
Comment (beiwei30) [inline dubbo-common/src/main/java/org/apache/dubbo/common/Constants.java]: rename to 'protobuf-json'

File: dubbo-dependencies-bom/pom.xml [hunk 4]
```diff
@@ -108,6 +108,8 @@
         <cxf_version>3.1.15</cxf_version>
         <thrift_version>0.12.0</thrift_version>
         <hessian_version>4.0.38</hessian_version>
+        <protobuf-java>3.6.0</protobuf-java>
```
Comment (beiwei30) [inline dubbo-dependencies-bom/pom.xml]: simply use one property for both protobuf-java and protobuf-java-util, and name it 'protobuf-java_version'

File: dubbo-distribution/pom.xml [hunk 5]
```diff
@@ -335,6 +335,11 @@
             <artifactId>dubbo-serialization-gson</artifactId>
             <version>${project.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.apache.dubbo</groupId>
+            <artifactId>dubbo-serialization-googlePb</artifactId>
```
Comment (beiwei30) [inline dubbo-distribution/pom.xml]: rename to 'dubbo-serialization-protobuf-json'

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/filter/GenericFilter.java [hunk 6]
```diff
@@ -107,6 +107,22 @@ public Result invoke(Invoker<?> invoker, Invocation inv) throws RpcException {
                                             args[i].getClass().getName());
                         }
                     }
+                } else if (ProtocolUtils.isProtobufGenericSerialization(generic)) {
```
Comment (beiwei30) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/filter/GenericFilter.java]: make sure line width doesn't exceed 120 columns.

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/filter/GenericFilter.java [hunk 7]
```diff
@@ -107,6 +107,22 @@ public Result invoke(Invoker<?> invoker, Invocation inv) throws RpcException {
...
+                        try (UnsafeByteArrayInputStream is = new UnsafeByteArrayInputStream(((String) args[0]).getBytes())) {
+                            args[0] = ExtensionLoader.getExtensionLoader(Serialization.class)
+                                    .getExtension("" + Constants.GENERIC_SERIALIZATION_PROTOBUF)
+                                    .deserialize(null, is).readObject(method.getParameterTypes()[0]);
+                        } catch (Exception e) {
+                            throw new RpcException("Deserialize argument failed.", e);
+                        }
+                    } else {
+                        throw new RpcException(
+                                new StringBuilder("Generic serialization [").append(Constants.GENERIC_SERIALIZATION_PROTOBUF)
```
Comment (beiwei30) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/filter/GenericFilter.java]: no necessary to use StringBuilder I think.

**Comment by beiwei30** [comment]
pls. also revert your change in dubbo-distrbution/pom.xml. It is unnecessary any longer.

File: dubbo-serialization/dubbo-serialization-googlePb/src/main/resources/META-INF/dubbo/internal/org.apache.dubbo.common.serialize.Serialization [hunk 8]
```diff
@@ -0,0 +1 @@
+gernericprotobuf=org.apache.dubbo.common.serialize.protobuf.support.GenericProtobufSerialization
```
Comment (beiwei30) [inline dubbo-serialization/dubbo-serialization-googlePb/src/main/resources/META-INF/dubbo/internal/org.apache.dubbo.common.serialize.Serialization]: use 'protobuf-json'

File: dubbo-bom/pom.xml (hunk 2 above)
Comment (vio-lin) [reply dubbo-bom/pom.xml]: OK

File: dubbo-common/src/main/java/org/apache/dubbo/common/Constants.java (hunk 3 above)
Comment (vio-lin) [reply dubbo-common/src/main/java/org/apache/dubbo/common/Constants.java]: OK

File: dubbo-dependencies-bom/pom.xml (hunk 4 above)
Comment (vio-lin) [reply dubbo-dependencies-bom/pom.xml]: OK

File: dubbo-all/pom.xml (hunk 1 above)
Comment (vio-lin) [reply dubbo-all/pom.xml]: done.

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/filter/GenericFilter.java (hunk 7 above)
Comment (vio-lin) [reply dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/filter/GenericFilter.java]: Done .  Use "string" + "string"

File: dubbo-serialization/dubbo-serialization-googlePb/src/main/resources/META-INF/dubbo/internal/org.apache.dubbo.common.serialize.Serialization (hunk 8 above)
Comment (vio-lin) [reply dubbo-serialization/dubbo-serialization-googlePb/src/main/resources/META-INF/dubbo/internal/org.apache.dubbo.common.serialize.Serialization]: Done

**Comment by beiwei30** [comment]
LGTM now.

**Comment by beiwei30** [comment]
there's one CI failure:

```sh
[ERROR]   The project org.apache.dubbo:dubbo-serialization:2.7.2-SNAPSHOT (/home/travis/build/apache/incubator-dubbo/dubbo-serialization/pom.xml) has 1 error
[ERROR]     Child module /home/travis/build/apache/incubator-dubbo/dubbo-serialization/dubbo-serialization-protobuf-json of /home/travis/build/apache/incubator-dubbo/dubbo-serialization/pom.xml does not exist
[ERROR]
```

**Comment by chickenlj** [comment]
This is a quite dependent module, I think we can merge it to the main branch and keep improving it. I will give it a test once I get some time.

**Pull request merged**